source/common/router/scoped_config_impl.cc

```diff
@@ -16,65 +18,107 @@ bool ScopeKey::operator==(const ScopeKey& other) const {
   return this->hash() == other.hash();
 }
 
-HeaderValueExtractorImpl::HeaderValueExtractorImpl(
-    ScopedRoutes::ScopeKeyBuilder::FragmentBuilder&& config)
-    : FragmentBuilderBase(std::move(config)),
-      header_value_extractor_config_(config_.header_value_extractor()) {
-  ASSERT(config_.type_case() ==
-             ScopedRoutes::ScopeKeyBuilder::FragmentBuilder::kHeaderValueExtractor,
-         "header_value_extractor is not set.");
-  if (header_value_extractor_config_.extract_type_case() ==
-      ScopedRoutes::ScopeKeyBuilder::FragmentBuilder::HeaderValueExtractor::kIndex) {
-    if (header_value_extractor_config_.index() != 0 &&
-        header_value_extractor_config_.element_separator().empty()) {
-      throw ProtoValidationException("Index > 0 for empty string element separator.",
-                                     header_value_extractor_config_);
+void FragmentBuilderImpl::validateHeaderValueExtractorConfig(
+    const HeaderValueExtractorConfig& config) const {
+  if (config.extract_type_case() == HeaderValueExtractorConfig::kIndex) {
+    if (config.index() != 0 && config.element_separator().empty()) {
+      throw ProtoValidationException("Index > 0 for empty string element separator.", config);
     }
   }
-  if (header_value_extractor_config_.extract_type_case() ==
-      ScopedRoutes::ScopeKeyBuilder::FragmentBuilder::HeaderValueExtractor::EXTRACT_TYPE_NOT_SET) {
-    throw ProtoValidationException("HeaderValueExtractor extract_type not set.",
-                                   header_value_extractor_config_);
+
+  if (config.extract_type_case() == HeaderValueExtractorConfig::EXTRACT_TYPE_NOT_SET) {
+    throw ProtoValidationException("HeaderValueExtractor extract_type not set.", config);
+  }
+}
+
+void FragmentBuilderImpl::validateMetadataValueExtractorConfig(
+    const MetadataValueExtractorConfig& config) const {
+  // Assuming support for only a single type of metadata extraction.
+  ASSERT(config.metadata_extract_type_case() == MetadataValueExtractorConfig::kMetadataKey);
+}
+
+FragmentBuilderImpl::FragmentBuilderImpl(ScopedRoutes::ScopeKeyBuilder::FragmentBuilder config)
```

Comment:
A drive-by question: why is `config`  not rvalue now? This ctor seems to be called only once with `ScopedRoutes::ScopeKeyBuilder::FragmentBuilder` instantiated at the call site.

Reply:
You're right, but it's still an rvalue. It's just not forced on the callers.